src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldNotHaveIncorrectSuffix.cs

```diff
@@ -56,13 +61,182 @@ public abstract class IdentifiersShouldNotHaveIncorrectSuffixAnalyzer : Diagnost
+                        return;
+                    }
+
+                    string name = namedTypeSymbol.Name;
+                    Compilation compilation = context.Compilation;
+
+                    foreach (string suffix in s_suffixToTypeDictionary.Keys)
+                    {
+                        if (IsNotChildButHasSuffix(
+                            namedTypeSymbol,
+                            compilation.GetTypeByMetadataName(s_suffixToTypeDictionary[suffix]),
+                            suffix,
+                            compilation))
+                        {
+                            context.ReportDiagnostic(
+                                namedTypeSymbol.CreateDiagnostic(TypeNoAlternateRule, name, suffix));
+                            return;
+                        }
+                    }
+
+                    foreach (string suffix in s_suffixToAllowedTypesDictionary.Keys)
+                    {
+                        if (name.HasSuffix(suffix)
+                            && !s_suffixToAllowedTypesDictionary[suffix].Contains(name))
+                        {
+                            context.ReportDiagnostic(
+                                namedTypeSymbol.CreateDiagnostic(TypeNoAlternateRule, name, suffix));
+                            return;
+                        }
+                    }
+
+                    if (name.HasSuffix(ImplSuffix))
+                    {
+                        context.ReportDiagnostic(
+                            namedTypeSymbol.CreateDiagnostic(MemberWithAlternateRule, ImplSuffix, name, CoreSuffix));
+                        return;
+                    }
+
+                    // FxCop performed the length check for "Ex", but not for any of the other
+                    // suffixes, because alone among the suffixes, "Ex" is the only one that
+                    // isn't itself a known type or a language keyword.
+                    if (name.HasSuffix(ExSuffix) && name.Length > ExSuffix.Length)
+                    {
+                        context.ReportDiagnostic(
+                            namedTypeSymbol.CreateDiagnostic(TypeNewerVersionRule, ExSuffix, name));
+                        return;
+                    }
+
+                    if (name.HasSuffix(NewSuffix))
+                    {
+                        context.ReportDiagnostic(
+                            namedTypeSymbol.CreateDiagnostic(TypeNewerVersionRule, NewSuffix, name));
+                        return;
+                    }
+                }, SymbolKind.NamedType);
+
+            analysisContext.RegisterSymbolAction(
+                (SymbolAnalysisContext context) =>
+                {
+                    var methodSymbol = (IMethodSymbol)context.Symbol;
+                    if (methodSymbol.GetResultantVisibility() != SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    if (methodSymbol.IsOverride)
```

Comment:
FXCop does a bunch of additional checks in its IsOverride helper method. See [here](http://index/#FxCopSdk/RuleUtilities.cs,597). We seem to be doing only (1) here. You should probably also bail out if `IMethodSymbolExtensions.IsImplementationOfAnyInterfaceMethod` (defined in our AnalyzerUtilities) returns true to match FXCop.